File: src/lib/autostart.ts
import { useEffect, useState } from "react";
import { invoke } from "@tauri-apps/api/core";

/**
 * Autostart ("Bei Anmeldung starten").
 *
 * Die Quelle der Wahrheit ist das Betriebssystem (LaunchAgent auf macOS,
 * Registry-Run-Eintrag auf Windows) — NICHT localStorage. Darum lesen wir den
 * echten Zustand beim Mount via Rust-Command und schalten ebenfalls über Rust.
 * Dadurch bleibt der Settings-Toggle konsistent mit dem Tray-Haken
 * „Bei Login starten".
 */
export function useAutostart() {
  const [enabled, setEnabled] = useState(false);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    let alive = true;
    invoke<boolean>("autostart_get")
      .then((v) => alive && setEnabled(v))
      .catch((e) => console.warn("[autostart] get failed", e))
      .finally(() => alive && setLoading(false));
    return () => {
      alive = false;
    };
  }, []);

  const set = async (v: boolean) => {
    setEnabled(v); // optimistisch
    try {
      const actual = await invoke<boolean>("autostart_set", { enabled: v });
      setEnabled(actual);
    } catch (e) {
      console.warn("[autostart] set failed", e);
      // Auf den real geltenden Zustand zurückrollen
      try {
        setEnabled(await invoke<boolean>("autostart_get"));
      } catch {
        setEnabled(!v);
      }
    }
  };

  return { enabled, setEnabled: set, loading };
}

File: src-tauri/src/autostart.rs
// Autostart („Bei Anmeldung starten") — die OS-Registrierung ist die Wahrheit,
// NICHT ein localStorage-Flag. macOS legt einen LaunchAgent an, Windows einen
// Registry-Run-Eintrag (via tauri-plugin-autostart). Settings-Toggle UND
// Tray-Haken „Bei Login starten" laufen über denselben Pfad, damit beide immer
// denselben Zustand zeigen.
use tauri::menu::CheckMenuItem;
use tauri::{Manager, Wry};
use tauri_plugin_autostart::ManagerExt;

/// Hält den Tray-Haken als managed State, damit `apply()` ihn synchron halten
/// kann. Konkreter Runtime `Wry` (gleicher Runtime wie der Rest der App), weil
/// `CheckMenuItem` keinen Default-Generic hat.
pub struct AutostartMenu(pub CheckMenuItem<Wry>);

/// Autostart setzen, Tray-Haken nachziehen und den danach tatsächlich
/// geltenden Zustand zurückgeben (OS bleibt Quelle der Wahrheit).
pub fn apply(app: &tauri::AppHandle, enabled: bool) -> Result<bool, String> {
    let mgr = app.autolaunch();
    if enabled { mgr.enable() } else { mgr.disable() }.map_err(|e| e.to_string())?;
    let now = mgr.is_enabled().unwrap_or(enabled);
    if let Some(menu) = app.try_state::<AutostartMenu>() {
        let _ = menu.0.set_checked(now);
    }
    Ok(now)
}

/// Aktueller Autostart-Zustand (vom Betriebssystem gelesen).
#[tauri::command]
pub fn autostart_get(app: tauri::AppHandle) -> bool {
    app.autolaunch().is_enabled().unwrap_or(false)
}

/// Autostart ein-/ausschalten. Gibt den real geltenden Zustand zurück, damit
/// das Frontend bei einem Fehlschlag den Toggle korrigieren kann.
#[tauri::command]
pub fn autostart_set(app: tauri::AppHandle, enabled: bool) -> Result<bool, String> {
    apply(&app, enabled)
}
